File: scripts/lineups_to_csv.rb
require "nokogiri"
require 'open-uri'
require 'date'
require_relative 'functions.rb'

all_the_matches = File.readlines("data/matches.csv").
                       map { |row| row.gsub(/\n/, "").split(",") }

MATCHES = {}

all_the_matches.each do |match|
	match_id = match[5].gsub(/\.html/, "").split("/")[-2].split("=")[1]
	MATCHES[match_id] = {
		:match_number => match[-1],
		:year => match[0],
		:home => { :starting => [], :subs => [] },
		:away => { :starting => [], :subs => [] }
	}
end

match_files = Dir["data/matches/*"]

match_files.select {|file| !file.end_with? ".md" }.each do |match_file|
	match_id = match_file.gsub(/\.html/, "").split("/")[-1]
	puts match_id
	year = MATCHES[match_id][:year]

	next if year == "2010"
	
	@doc = Nokogiri::HTML(open(match_file))
	world_cup = @doc.css("div.content_header h1").text
	
	home_team =  @doc.css("div.lnupTeam").first
	home_team.css("ul:nth-child(2) li").each do |row|
		player = row.css("div.playerInfo span").text
		MATCHES[match_id][:home][:starting] << Functions.clean_up_player(player)
	end
	
	home_team.css("ul:nth-child(3) li").each do |row|
		player = row.css("div.playerInfo span").text
		MATCHES[match_id][:home][:subs] << Functions.clean_up_player(player)
	end

	away_team = @doc.css("div.lnupTeam.away")
	away_team.css("ul:nth-child(2) li").each do |row|
		player = row.css("div.playerInfo span").text
		MATCHES[match_id][:away][:starting] << Functions.clean_up_player(player)
	end
	
	away_team.css("ul:nth-child(3) li").each do |row|
		player = row.css("div.playerInfo span").text
		MATCHES[match_id][:away][:subs] << Functions.clean_up_player(player)
	end	

	MATCHES[match_id] = MATCHES[match_id].merge( {:world_cup => world_cup})
end

open("data/import/lineups.csv", 'w') { |f|	
	f.puts (["world_cup",
		     "match_id", 
			 "team",
			 "type",
			 "player",
			 ""]).join(",")

  	MATCHES.each do |match_id, row|
  		row[:home][:starting].each do |player|
			f.puts ["\"#{row[:world_cup]}\"",
			        "\"#{match_id}\"",
			    	"\"home\"",
			   		"\"starting\"" ,
					"\"#{player}\""
					].join(",")
  		end

  		row[:home][:subs].each do |player|
			f.puts ["\"#{row[:world_cup]}\"",
			        "\"#{match_id}\"",
			    	"\"home\"",
			   		"\"sub\"" ,
					"\"#{player}\""
					].join(",")
  		end  

  		row[:away][:starting].each do |player|
			f.puts ["\"#{row[:world_cup]}\"",
			        "\"#{match_id}\"",
			    	"\"away\"",
			   		"\"starting\"" ,
					"\"#{player}\""
					].join(",")
  		end

  		row[:away][:subs].each do |player|
			f.puts ["\"#{row[:world_cup]}\"",
			        "\"#{match_id}\"",
			    	"\"away\"",
			   		"\"sub\"" ,
					"\"#{player}\""
					].join(",")
  		end 
	end
}

File: scripts/python/lineups_to_csv.py
import csv
import glob
import json

import re
from bs4 import BeautifulSoup

with open("data/2018/import/lineups.csv", "w") as lineups_file:
    writer = csv.writer(lineups_file, delimiter=",")

    writer.writerow([
        "year", "match_id", "team_id", "type", "player_id"
    ])

    for match_file in glob.glob("data/2018/matches/*.html"):
        match_id = match_file.split("/")[-1].replace(".html", "")
        print(match_id)
        with open(match_file, "r") as file:
            soup = BeautifulSoup(file.read(), "html.parser")

            header_element = soup.select("div h1 a")

            if len(header_element) == 0:
                header_element = soup.select("div a h1")

            header = header_element[0].text
            groups = re.match("(\d{4}) FIFA World Cup ([a-zA-Z ]*)", header)
            year = int(groups[1])

            if year == 2018:
                with open("data/2018/matches/api/{0}.json".format(match_id)) as json_file:
                    the_json = json.loads(json_file.read())

                    home_team = the_json["HomeTeam"]
                    players = home_team["Players"]
                    for player in players:
                        type = "STARTED" if player["Status"] == 1 else "SUBSTITUTE"
                        writer.writerow([
                            year, match_id, home_team["IdTeam"], type, player["IdPlayer"]
                        ])

                    away_team = the_json["HomeTeam"]
                    players = away_team["Players"]
                    for player in players:
                        type = "STARTED" if player["Status"] == 1 else "SUBSTITUTE"
                        writer.writerow([
                            year, match_id, away_team["IdTeam"], type, player["IdPlayer"]
                        ])

            else:
                teams = [team["data-team-id"] for team in soup.select("td.team-name div.t")]

                home_starters = soup.select("table.fielded tr td.home div.p-i-no")
                for player in home_starters:
                    writer.writerow([
                        year, match_id, teams[0], "STARTED", player["data-player-id"]
                    ])

                away_starters = soup.select("table.fielded tr td.away div.p-i-no")
                for player in away_starters:
                    writer.writerow([
                        year, match_id, teams[1], "STARTED", player["data-player-id"]
                    ])

                home_subs = soup.select("table.substitutes tr td.home div.p-i-no")
                for player in home_subs:
                    writer.writerow([
                        year, match_id, teams[0], "SUBSTITUTE", player["data-player-id"]
                    ])

                away_subs = soup.select("table.substitutes tr td.away div.p-i-no")
                for player in away_subs:
                    writer.writerow([
                        year, match_id, teams[1], "SUBSTITUTE", player["data-player-id"]
                    ])